### PR TITLE
Add DataTable Events

### PR DESCRIPTION
**Pull request by MaximePinot**
Hi,

It can be useful to be able to modify the Query before the iteration starts. I personally need to do this so I can configure the Doctrine cache.

Here's how it works:

```php?start_inline=1
$table->createAdapter(ORMAdapter::class, [
    'entity' => Employee::class,
    'query' => function (QueryBuilder $builder) {
        $builder
            ->select('e')
            ->addSelect('c')
            ->from(Employee::class, 'e')
            ->leftJoin('e.company', 'c')
        ;
    },
]);

$table->addEventListener(ORMAdapterEvents::PRE_QUERY, function(ORMAdapterQueryEvent $event) {
    $event->getQuery()->useResultCache(true)->useQueryCache(true);
});
```

### CODE REVIEW

File: tests/Fixtures/services.yml [hunk 1]
```diff
@@ -2,6 +2,18 @@ services:
     Tests\Fixtures\AppBundle\DataTable\Type\ServicePersonTableType:
         autowire: true
         autoconfigure: true
+
+    doctrine.result_cache_provider:
+        class: Symfony\Component\Cache\DoctrineProvider
+        public: false
+        arguments:
+            - '@doctrine.result_cache_pool'
+    doctrine.system_cache_provider:
```
Comment (curry684) [inline tests/Fixtures/services.yml]: Newline between service definitions

File: src/Adapter/Doctrine/ORMAdapterEvents.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+     * The PRE_QUERY event is dispatched after the QueryBuilder
+     * built the Query and before the iteration starts.
+     *
+     * It can be useful to configure the cache.
+     *
+     * @Event("Omines\DataTablesBundle\Adapter\Doctrine\Event\ORMAdapterQueryEvent")
+     */
+    const PRE_QUERY = 'omines_datatables.pre_query';
+
+    private function __construct()
```
Comment (curry684) [inline src/Adapter/Doctrine/ORMAdapterEvents.php]: Constructor is not needed as it is empty and the class won't be initialized.

File: src/Adapter/Doctrine/ORMAdapter.php [hunk 3]
```diff
@@ -54,18 +56,24 @@ class ORMAdapter extends AbstractAdapter
...
+    protected $eventDispatcher;
+
     /**
      * DoctrineAdapter constructor.
      *
+     * @param EventDispatcherInterface $eventDispatcher
      * @param RegistryInterface|null $registry
      */
-    public function __construct(RegistryInterface $registry = null)
+    public function __construct(EventDispatcherInterface $eventDispatcher, RegistryInterface $registry = null)
```
Comment (curry684) [inline src/Adapter/Doctrine/ORMAdapter.php]: The adapter does not need to get passed the dispatcher.

File: src/Adapter/Doctrine/ORMAdapter.php [hunk 4]
```diff
@@ -197,8 +205,11 @@ protected function getResults(AdapterQuery $query): \Traversable
             ;
         }
 
-        //!$this->query->getHint(self::HINT_DISTINCT) || isset($this->selectedClasses[$joinedDqlAlias])
-        foreach ($builder->getQuery()->iterate([], $this->hydrationMode) as $result) {
+        $query = $builder->getQuery();
+        $event = new ORMAdapterQueryEvent($query);
+        $this->eventDispatcher->dispatch(ORMAdapterEvents::PRE_QUERY, $event);
```
Comment (curry684) [inline src/Adapter/Doctrine/ORMAdapter.php]: You can access the dispatcher via `$state->getTable()`, that's why the adapter does not need to be aware of the EventDispatcher.

File: src/DataTable.php [hunk 5]
```diff
@@ -65,6 +66,9 @@ class DataTable
     /** @var array<string, AbstractColumn> */
     protected $columnsByName = [];
 
+    /** @var EventDispatcherInterface */
+    protected $eventDispatcher;
```
Comment (curry684) [inline src/DataTable.php]: Needs a getter.

**Comment by curry684** [comment]
The core change we just requested is about keeping 1 central reference to the event dispatcher. Passing it around to the adapters may cause issues in the future.

File: src/Adapter/Doctrine/ORMAdapterEvents.php (hunk 2 above)
Comment (curry684) [reply src/Adapter/Doctrine/ORMAdapterEvents.php]: Ref https://github.com/symfony/http-kernel/blob/master/KernelEvents.php

File: src/Adapter/Doctrine/ORMAdapterEvents.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /**
+     * The PRE_QUERY event is dispatched after the QueryBuilder
+     * built the Query and before the iteration starts.
+     *
+     * It can be useful to configure the cache.
+     *
+     * @Event("Omines\DataTablesBundle\Adapter\Doctrine\Event\ORMAdapterQueryEvent")
+     */
+    const PRE_QUERY = 'omines_datatables.pre_query';
```
Comment (curry684) [inline src/Adapter/Doctrine/ORMAdapterEvents.php]: Should be namespaced to the adapter: `omines_datatables.ormadapter.pre_query`.

**Comment by MaximePinot** [comment]
Thank you for your feedback @curry684! Your requested changes are resolved in 04580e7.

**Comment by curry684** [comment]
Thank you @MaximePinot!

**Pull request merged**